### PR TITLE
Support S3 namespaces when retrieving models from buckets

### LINKED ISSUE #4916: Models loaded from S3 buckets do not support namespaces
**Rasa version**: 1.5.0

**Rasa SDK version** (if used & relevant): n/a

**Rasa X version** (if used & relevant): n/a

**Python version**:3.7.4

**Operating system** (windows, osx, ...): osx & linux

**Issue**: When working with many models, it helps to make use of S3 namespacing functionality (aka "folders") within buckets. Not being able to load models from buckets is a hindrance when working across projects. 
Models can only be loaded from the root directory of an S3 bucket. This is because when the full S3 path passed in the -m flag is then passed, the full path is where the model is attempted to write to disk. 
This full path is also used to attempt to decompress the model.

**Error (including full traceback)**:
```
2019-12-06 09:04:53 ERROR    rasa.core.agent  - Could not load model due to [Errno 2] No such file or directory: 'projectA/20191205-192442.tar.gz'.
Experienced exception while trying to serve
Traceback (most recent call last):
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/app.py", line 1096, in run
    serve(**server_settings)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/server.py", line 742, in serve
    trigger_events(before_start, loop)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/server.py", line 604, in trigger_events
    loop.run_until_complete(result)
  File "uvloop/loop.pyx", line 1417, in uvloop.loop.Loop.run_until_complete
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/run.py", line 244, in load_agent_on_start
    action_endpoint=endpoints.action,
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/agent.py", line 249, in load_agent
    model_server=model_server,
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/agent.py", line 932, in load_from_remote_storage
    persistor.retrieve(model_name, target_path)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/nlu/persistor.py", line 55, in retrieve
    self._retrieve_tar(tar_name)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/nlu/persistor.py", line 157, in _retrieve_tar
    with io.open(target_filename, "wb") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'projectA/20191205-192442.tar.gz'
Traceback (most recent call last):
  File "/Users/cianclarke/Library/Python/3.7/bin/rasa", line 8, in <module>
    sys.exit(main())
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/__main__.py", line 76, in main
    cmdline_arguments.func(cmdline_arguments)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/cli/run.py", line 100, in run
    rasa.run(**vars(args))
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/run.py", line 56, in run
    **kwargs
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/run.py", line 202, in serve_application
    backlog=int(os.environ.get("SANIC_BACKLOG", "100")),
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/app.py", line 1096, in run
    serve(**server_settings)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/server.py", line 742, in serve
    trigger_events(before_start, loop)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/sanic/server.py", line 604, in trigger_events
    loop.run_until_complete(result)
  File "uvloop/loop.pyx", line 1417, in uvloop.loop.Loop.run_until_complete
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/run.py", line 244, in load_agent_on_start
    action_endpoint=endpoints.action,
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/agent.py", line 249, in load_agent
    model_server=model_server,
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/core/agent.py", line 932, in load_from_remote_storage
    persistor.retrieve(model_name, target_path)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/nlu/persistor.py", line 55, in retrieve
    self._retrieve_tar(tar_name)
  File "/Users/cianclarke/Library/Python/3.7/lib/python/site-packages/rasa/nlu/persistor.py", line 157, in _retrieve_tar
    with io.open(target_filename, "wb") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'projectA/20191205-192442.tar.gz'

```

**Command or request that led to error**:
```
rasa run -m "projectA/20191205-192442.tar.gz" -p 5005 -vv --remote-storage=aws
```

A PR fixing this issue to follow.

### PR DESCRIPTION
**Pull request by cianclarke**
**Proposed changes**:
- When working with many models, it helps to make use of S3 namespacing functionality (aka "folders") within buckets. Not being able to load models from buckets is a hindrance when working across projects. 
- Present S3 bucket support only works with root directory - whenever a namespace is provided, this path is also used on the local file system to write the `.tar.gz` (see #4916 )
- PR allows passing of S3 paths with namespaces for modal retrieval. 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

Fixes #4916

### CODE REVIEW

**Comment by erohmensing** [comment]
Thanks for the PR, we'll give it a review as soon as possible.

File: rasa/nlu/persistor.py [hunk 1]
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
         with open(tar_path, "rb") as f:
             self.s3.Object(self.bucket_name, file_key).put(Body=f)
 
-    def _retrieve_tar(self, target_filename: Text) -> None:
+    def _retrieve_tar(self, model_full_path: Text) -> None:
+        tar_name = os.path.basename(model_full_path)
```
Comment (ricwo) [inline rasa/nlu/persistor.py]: this line should move one down (below the docstring)

File: tests/nlu/base/test_persistor.py [hunk 2]
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
...
+        decompress.assert_called_once_with("model.tar.gz", destination)
+        retrieve.assert_called_once_with(model)
+
+
+# noinspection PyPep8Naming
+@mock_s3
+def test_s3_private_retrieve_tar():
+    model = "/my/s3/project/model.tar.gz"
+    awsPersistor = persistor.AWSPersistor("rasa-test")
+    with patch.object(awsPersistor.bucket, "download_fileobj") as download_fileobj:
```
Comment (ricwo) [inline tests/nlu/base/test_persistor.py]:
```suggestion
    with patch.object(awsPersistor.bucket, "download_fileobj") as download_fileobj:
        # noinspection PyProtectedMember
```

File: tests/nlu/base/test_persistor.py [hunk 3]
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
...
+        awsPersistor._retrieve_tar(model)
+    retrieveArgs = download_fileobj.call_args[0]
+    assert retrieveArgs[0] == model
+    assert retrieveArgs[1].name == "model.tar.gz"
```
Comment (ricwo) [inline tests/nlu/base/test_persistor.py]: can we move these two tests right below the existing two s3 tests?

File: tests/nlu/base/test_persistor.py [hunk 4]
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
...
+    with patch.object(persistor.AWSPersistor, "_decompress") as decompress:
+        with patch.object(persistor.AWSPersistor, "_retrieve_tar") as retrieve:
+            persistor.AWSPersistor("rasa-test").retrieve(model, destination)
+        decompress.assert_called_once_with("model.tar.gz", destination)
+        retrieve.assert_called_once_with(model)
+
+
+# noinspection PyPep8Naming
+@mock_s3
+def test_s3_private_retrieve_tar():
```
Comment (ricwo) [inline tests/nlu/base/test_persistor.py]: what exactly is being tested here? A comment would be very helpful

File: rasa/nlu/persistor.py (hunk 1 above)
Comment (cianclarke) [reply rasa/nlu/persistor.py]: Done!

File: tests/nlu/base/test_persistor.py (hunk 3 above)
Comment (cianclarke) [reply tests/nlu/base/test_persistor.py]: Done!

File: tests/nlu/base/test_persistor.py (hunk 4 above)
Comment (cianclarke) [reply tests/nlu/base/test_persistor.py]: Added! I included this test retroactively to bump coverage back to an acceptable level by coveralls. I'd be happy to also remove it if you think it excessive

**Comment by cianclarke** [comment]
@ricwo Many thanks for the review. I've incorporated all your suggestions, and also added a changelog entry 👍

File: rasa/nlu/persistor.py [hunk 5]
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
         with open(tar_path, "rb") as f:
             self.s3.Object(self.bucket_name, file_key).put(Body=f)
 
-    def _retrieve_tar(self, target_filename: Text) -> None:
+    def _retrieve_tar(self, model_full_path: Text) -> None:
```
Comment (ricwo) [inline rasa/nlu/persistor.py]:
```suggestion
    def _retrieve_tar(self, model_path: Text) -> None:
```

File: rasa/nlu/persistor.py [hunk 6]
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
...
         """Downloads a model that has previously been persisted to s3."""
-
-        with open(target_filename, "wb") as f:
-            self.bucket.download_fileobj(target_filename, f)
+        tar_name = os.path.basename(model_full_path)
```
Comment (ricwo) [inline rasa/nlu/persistor.py]:
```suggestion
        tar_name = os.path.basename(model_path)
```

File: rasa/nlu/persistor.py [hunk 7]
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
...
+        with open(tar_name, "wb") as f:
+            self.bucket.download_fileobj(model_full_path, f)
```
Comment (ricwo) [inline rasa/nlu/persistor.py]:
```suggestion
            self.bucket.download_fileobj(model_path, f)
```

File: rasa/nlu/persistor.py (hunk 5 above)
Comment (ricwo) [reply rasa/nlu/persistor.py]: in line with `model_path` args used elsewhere

**Comment by ricwo** [comment]
@cianclarke looks like you didn't check in the changelog file - would you mind adding that?

**Comment by cianclarke** [comment]
Ah, sorry @ricwo - you're absolutely right. I've committed the changelog entry. 
I'm also noticing coveralls oddly report coverage having gone back down again, just from moving the tests around when I addressed your PR feedback. I'm not sure what's going on there!

File: rasa/nlu/persistor.py (hunk 5 above)
Comment (ricwo) [reply rasa/nlu/persistor.py]: @cianclarke would you mind updating the args?

**Comment by cianclarke** [comment]
@ricwo - I committed your args suggestions from the GitHub UI (looks like I may have missed those). 
Hoping that was the argument updates you had meant - happy to also rebase the excess commits out, or can do on merge - not sure if the project has a preference on this.

**Comment by cianclarke** [comment]
Thanks @ricwo

**Pull request merged**